src/main/kotlin/org/rust/ide/inspections/RsDeprecationInspection.kt

```diff
@@ -0,0 +1,94 @@
+import com.intellij.psi.PsiElementVisitor
+import com.intellij.psi.util.parents
+import org.rust.lang.core.psi.RsFile
+import org.rust.lang.core.psi.RsMetaItem
+import org.rust.lang.core.psi.RsModDeclItem
+import org.rust.lang.core.psi.RsVisitor
+import org.rust.lang.core.psi.ext.*
+
+class RsDeprecationInspection : RsLocalInspectionTool() {
+    override fun getDisplayName() = "Deprecated item"
+
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor = object : RsVisitor() {
+        override fun visitElement(ref: RsElement) {
+            // item is non-inline module declaration or not reference element
+            if (ref is RsModDeclItem || ref !is RsWeakReferenceElement) return
+
+            val original = ref.reference?.resolve() ?: return
+            val identifier = ref.referenceNameElement ?: return
+
+            val targetElement = when {
+                original is RsFile -> original.declaration
+                original is RsAbstractable && original.owner.isImplOrTrait -> original.superItem ?: original
+                else -> original
+            } ?: return
+
+            if (ref.parents().none { it.hasAllowDeprecatedAttribute() }) {
+                checkAndRegisterAsDeprecated(identifier, targetElement, holder)
+            }
+        }
+    }
+
+    private fun checkAndRegisterAsDeprecated(identifier: PsiElement, original: PsiElement, holder: ProblemsHolder) {
+        if (original is RsOuterAttributeOwner) {
+            QueryAttributes(original)
+                .deprecatedAttribute
+                ?.let { DeprecatedAttribute.from(it) }
+                ?.let { holder.registerProblem(identifier, it.getMessage(identifier.text), LIKE_DEPRECATED) }
+        }
+    }
+
+    private fun PsiElement.hasAllowDeprecatedAttribute(): Boolean = this is RsDocAndAttributeOwner &&
+        QueryAttributes(this).hasAttributeWithArg(ALLOW_ATTR_NAME, DEPRECATED_ARG_NAME)
+
+    companion object {
+        private const val ALLOW_ATTR_NAME: String = "allow"
+        private const val DEPRECATED_ARG_NAME: String = "deprecated"
+    }
+}
+
+private class DeprecatedAttribute(val note: String?, val since: String?) {
+
+    fun getMessage(item: String): String = buildString {
+        append("`$item` is deprecated")
+        if (since != null) append(" since $since")
+        if (note != null) append(": $note")
+    }
+
+    companion object {
+        private const val DEPRECATED_ATTR_NAME: String = "deprecated"
+
+        private const val SINCE_PARAM_NAME: String = "since"
+        private const val NOTE_PARAM_NAME: String = "note"
+        private const val REASON_PARAM_NAME: String = "reason"
+
+        // #[deprecated] or #[rustc_deprecated]
+        fun from(deprecatedMetaItem: RsMetaItem): DeprecatedAttribute {
+            fun List<RsMetaItem>.getByName(name: String): String? =
+                firstOrNull { name == it.identifier?.text }?.litExpr?.stringLiteralValue
```

Comment:
Let's use `RsMetaItem.name` and `RsMetaItem.value` extension properties here instead of `it.identifier?.text` and `litExpr?.stringLiteralValue` not only that we already have such code but also because they use stubs if it's possible, i.e. they don't reparse not opened files
Also, it would be cool to add a test to check that we don't load ast for all files
You need to create a multifile test where a deprecated item is located not in a file where you are using and just call `org.rust.RsTestBase#checkAstNotLoaded` before inspection run